File: Entity.java
// Entity class

// This class holds data on ONE entity. It's only really useful when
// used in an array along with many others. Each value is stored as
// a separate attribute, in an array.

// A small note, I was tempted to use an Attribute class and just
// make an array of that, but that's breaking it down too far.

import java.util.Scanner; // Perfect for String handling
import java.io.FileInputStream;
import java.io.FileOutputStream;

public class Entity {
	
	// INITIAL DATA DECLARATION AND DEFINITION OF CONSTANTS
	
	private String[] attributes;
	private int numAttributes=0;
	
	public static int X=0;
	public static int Y=1;
	public static int Z=2;
	
	// CONSTRUCTORS
	
	public Entity(byte[] in) {
		String me="";
		for(int i=0;i<in.length;i++) {
			me+=(char)in[i];
		}
		setData(me);
	}

	public Entity(String me) {
		setData(me);
	}
	
	// METHODS
	
	// deleteAttribute(String)
	// Takes: String (attribute to delete)
	// Returns: Success boolean (in case the attribute wasn't found)
	// Deletes the specified attribute from the attributes list. If it wasn't found it does nothing. If there
	// are duplicates this only deletes the first one. Just while loop it. while(deleteAttribute("Attribute");
	public boolean deleteAttribute(String attribute) {
		int index=0;
		boolean found=false;
		for(index=0;index<numAttributes;index++) {
			try {
				if(attributes[index].substring(0,attribute.length()+2).compareToIgnoreCase("\""+attribute+"\"")==0) {
					found=true;
					break; // If the attribute is found, break the loop since the appropriate index in in the variable "index"
				}
			} catch(StringIndexOutOfBoundsException e) { // for cases where the whole String is shorter than
				;                                         // the name of the attribute we're looking for. Do nothing.
			}
		}
		if (found) {
			deleteAttribute(index);
			return true;
		}
		return false;
	}
	
	// deleteAllOfAttribute(String)
	// Deletes all instances of specified attribute.
	public void deleteAllOfAttribute(String attribute) {
		while(deleteAttribute(attribute));
	}
	
	// deleteAttribute(int)
	// Deletes the attribute at the specified index in the list
	public void deleteAttribute(int index) {
		String[] newList=new String[numAttributes-1];
		for(int i=0;i<numAttributes-1;i++) {
			if(i<index) {
				newList[i]=attributes[i];
			}
			if(i>=index) {
				newList[i]=attributes[i+1];
			}
		}
		attributes=newList;
		numAttributes--;
	}
	
	// addAttribute(String)
	// Simply adds the input String to the attribute list. This String can be anything,
	// even containing newlines or curly braces. BE CAREFUL.
	public void addAttribute(String in) {
		String[] newList=new String[numAttributes+1];
		for(int i=0;i<numAttributes;i++) { // copy the current attribute list
			newList[i]=attributes[i];
		}
		newList[numAttributes]=in;
		attributes=newList;
		numAttributes++;
	}
	
	// addAttribute(String, String)
	// Adds the specified attribute with the specified value to the list.
	public void addAttribute(String attribute, String value) {
		addAttribute("\""+attribute+"\" \""+value+"\"");
	}
	
	// +toByteArray()
	// Returns the entity as an array of bytes. This could be faster than just
	// concatenating all the damn Strings, it depends on how the String class
	// handles the getBytes() conversion. Judging by how much faster my code
	// is than the stupid DataOutputStream, it won't be great. I might have
	// to look at the code in the String class and try to think of a better
	// way than what they have there.
	//
	// This works much faster than toString, especially for entities with a LOT
	// of data. All the String concatenation in toString was building up, with
	// thousands of Strings to stick together and
	public byte[] toByteArray() {
		byte[] out;
		int len=0;
		// Get the lengths of all attributes together
		for(int i=0;i<attributes.length;i++) {
			len+=attributes[i].length()+1; // Gonna need a newline after each attribute or they'll get jumbled together
		}
		out=new byte[len];
		int offset=0;
		for(int i=0;i<attributes.length;i++) { // For each attribute
			for(int j=0;j<attributes[i].length();j++) { // Then for each byte in the attribute
				out[j+offset]=(byte)attributes[i].charAt(j); // add it to the output array
			}
			out[offset+attributes[i].length()]=(byte)0x0A;
			offset+=attributes[i].length()+1;
		}
		return out;
	}

	// +toString()
	// Returns the entity as an ASCII entity structure. The output of this method
	// reads as a complete entity that could be put into a map with no problems,
	// and that will be the primary use case for this method. Be sure to add the
	// newlines and curly braces around the entity if using this method to create
	// a new entities lump file, unless curly braces are part of the attributes
	// array.
	public String toString() {
		String out="";
		for(int i=0;i<attributes.length;i++) {
			if(!(attributes[i].charAt(0)=='}')) {
				out+=attributes[i]+"\n";
			} else { // Character is '}'
				out+=attributes[i];
			}
		}
		return out;
	}
	
	// +isBrushBased()
	// Reads the first character of the model attribute. If it's *, then it's a brush
	// based entity and this method returns true. If not, it returns false.
	// The worldspawn entity is also considered brush-based since it's model can be assumed
	// to be 0.
	public boolean isBrushBased() {
		try {
			return ((getAttribute("model").charAt(0)=='*') || (getAttribute("classname").equalsIgnoreCase("worldspawn")));
		} catch(java.lang.StringIndexOutOfBoundsException e) { // The entity has no "model"
			return false;
		}
	}
	
	// +setData(String)
	// Used by constructors and can be used by outside classes to set or change the data
	// in this specific entity. It takes a String and parses it into the attributes array.
	// This input String CAN include the { and } from the entity structure, in that it 
	// won't cause any errors in the program. However if and when you want to write the
	// entities back into a lump you must remember whether you included them or not. The
	// behavior of the Lump00 class is not to include them.
	public void setData(String in) {
		Scanner reader=new Scanner(in);
		reader.useDelimiter((char)0x0A+"");
		Scanner counter=new Scanner(in);
		counter.useDelimiter((char)0x0A+"");
		
		numAttributes=0;
		while(counter.hasNext()) {
			counter.next();
			numAttributes++;
		}
		
		attributes=new String[numAttributes];
		for(int i=0;i<numAttributes;i++) {
			String current=reader.next();
			// This will trim all the 0D bytes before the 0A delimiters if they exist,
			// since the Windows newline sequence is 0D0A and all its text editors use
			// that. This keeps the data from getting confusing to this program, but
			// also saves a small amount of space in the output lump itself.
			if(current.charAt(current.length()-1)==(char)0x0D) {
				current=current.substring(0,current.length()-1);
			}
			attributes[i]=current;
		}
	}
	
	// +getAttribute(String)
	// Takes in an attribute as a String and returns the value of that attribute,
	// if it exists. If not, return null. I used to have an exception for this,
	// but always catching it was a pain in the ass. So instead, if the attribute
	// doesn't exist, just return an empty String. I don't think I've ever seen
	// an empty string used as a value in a map before, and either way the setAttribute
	// method will automatically add an attribute if it doesn't exist, and just
	// change it if it does, whether it's empty or not.
	public String getAttribute(String attribute) {
		String output="";
		for(int i=0;i<numAttributes;i++) {
			try {
				if(attributes[i].substring(0,attribute.length()+2).compareToIgnoreCase("\""+attribute+"\"")==0) {
					output=attributes[i].substring(attribute.length()+4,attributes[i].length()-1);
					break;
				}
			} catch(StringIndexOutOfBoundsException e) { // for cases where the whole String is shorter than
				;                                         // the name of the attribute we're looking for. Do nothing.
			}
		}
		return output;
	}
	
	// +getModelNumber()
	// If there's a model number in the attributes list, this method fetches it
	// and returns it. If there is no model defined, or it's not a numerical 
	// value, then -1 is returned.
	public int getModelNumber() {
		int number=-1;
		for(int i=0;i<numAttributes;i++) {
			try {
				if(attributes[i].substring(0,7).compareToIgnoreCase("\"model\"")==0) {
					// This substring skips the "model" "* and gets to the number
					number=Integer.parseInt(attributes[i].substring(10,attributes[i].length()-1));
					break;
				}
			} catch(StringIndexOutOfBoundsException e) { // substring(0,7) was longer than the String
				;
			} catch(NumberFormatException e) { // The model wasn't a number
				;
			}
		}
		return number;
	}
	
	// setAttribute()
	// Set an attribute. If it doesn't exist, it is added. If it does, it is
	// overwritten with the new one, since that's much easier to do than edit
	// the preexisting one. If "value" is empty the attribute is deleted.
	public void setAttribute(String attribute, String value) {
		boolean done=false;
		if(value.length()>0) {
			for(int i=0;i<numAttributes && !done;i++) {
				try {
					if(attributes[i].substring(0,attribute.length()+2).compareToIgnoreCase("\""+attribute+"\"")==0) {
						attributes[i]="\""+attribute+"\" \""+value+"\"";
						done=true;
					}
				} catch(StringIndexOutOfBoundsException e) {
					;
				}
			}
			if(!done) {
				addAttribute(attribute, value);
			}
		} else { // If value is empty
			deleteAllOfAttribute(attribute);
		}
	}
	
	// getOrigin()
	// Returns the three components of the entity's "origin" attribute as an array
	// of three doubles. Since everything is a string anyway, I can be as precise
	// as I want.
	public double[] getOrigin() {
		double[] output=new double[3]; // initializes to {0,0,0}
		if(!getAttribute("origin").equals("")) {
			String origin=getAttribute("origin");
			Scanner numGetter=new Scanner(origin);
			for(int i=0;i<3&&numGetter.hasNext();i++) { // If the origin attribute is empty (doesn't exist) this will do nothing
				output[i]=numGetter.nextDouble();
			}
		}
		return output; // If origin didn't exist, this returns (0,0,0)
	}
	
	// setOrigin()
	// Sets the three components of the entity's "origin" attribute to an array
	// of three doubles.
	public void setOrigin(double[] in) {
		String origin="";
		if(in.length>=3) {
			origin=in[X]+" "+in[Y]+" "+in[Z];
		} else {
			if(in.length==2) {
				origin=in[X]+" "+in[Y];
			} else {
				if(in.length==1) {
					origin=Double.toString(in[X]);
				}
			}
		}
		setAttribute("origin", origin);
	}
	
	// getAngles()
	// Returns the three components of the entity's "angles" attribute as an array
	// of three doubles.
	public double[] getAngles() {
		double[] output=new double[3]; // initializes to {0,0,0}
		if(!getAttribute("angles").equals("")) {
			String angles=getAttribute("angles");
			Scanner numGetter=new Scanner(angles);
			for(int i=0;i<3&&numGetter.hasNext();i++) { // If the angles attribute is empty (doesn't exist) this will do nothing
				output[i]=numGetter.nextDouble();
			}
		}
		return output; // If angles didn't exist, this returns (0,0,0)
	}
	
	// setAngles()
	// Sets the three components of the entity's "angles" attribute to an array
	// of three doubles.
	public void setAngles(double[] in) {
		String angles="";
		if(in.length>=3) {
			angles=in[X]+" "+in[Y]+" "+in[Z];
		} else {
			if(in.length==2) {
				angles=in[X]+" "+in[Y];
			} else {
				if(in.length==1) {
					angles=Double.toString(in[X]);
				}
			}
		}
		setAttribute("angles", angles);
	}
	
	// getNumAttributes()
	// Returns the number of attributes in the entity
	public int getNumAttributes() {
		return attributes.length;
	}
}

File: LumpTools/Entity.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
// Entity class

// This class holds data on ONE entity. It's only really useful when
// used in an array along with many others. Each value is stored as
// a separate attribute, in a dictionary.

// I've also added the ability to add MAPBrush objects to the entity,
// to be processed later on.
[Serializable]
public class Entity:LumpObject {
	
	// INITIAL DATA DECLARATION AND DEFINITION OF CONSTANTS
	
	private Dictionary<string, string> attributes = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
	private List<MAPBrush> brushes = new List<MAPBrush>();
	
	// For building more Source engine connections, a simple state machine.
	private bool fired = false;
	
	// CONSTRUCTORS
	
	public Entity(byte[] data):this(Encoding.ASCII.GetString(data).Split((char)0x0A)) {
	}
	
	public Entity(string classname):base(new byte[0]) {
		attributes.Add("classname", classname);
	}
	
	public Entity(string[] atts):base(new byte[0]) {
		attributes = new Dictionary<string, string>(atts.Length, StringComparer.InvariantCultureIgnoreCase);
		for (int i = 0; i < atts.Length; i++) {
			addAttribute(atts[i]);
		}
	}
	
	public Entity():base(new byte[0]) {
	}
	
	public Entity(Entity copy):base(copy.Data) {
		attributes = new Dictionary<string, string>(copy.Attributes);
		brushes = new List<MAPBrush>(copy.Brushes);
	}

	public Entity(Dictionary<string, string> atts, List<MAPBrush> brshs):base(new byte[0]) {
		attributes = new Dictionary<string,string>(atts);
		brushes = new List<MAPBrush>(brshs);
	}

	public Entity(Dictionary<string, string> atts):this(atts, new List<MAPBrush>()) {
	}
	
	// METHODS
	
	// renameAttribute(String, String)
	// Renames the specified attribute to the second String.
	public virtual void renameAttribute(string attribute, string to) {
		if(attributes.ContainsKey(attribute)) {
			string val = attributes[attribute];
			attributes.Remove(attribute);
			if(attributes.ContainsKey(to)) { attributes.Remove(to); Console.WriteLine("WARNING: Attribute "+to+" already existed in entity, overwritten!"); }
			attributes.Add(to, val);
		}
	}
	
	// addAttribute(String)
	// Simply adds the input String to the attribute list. This String can be anything,
	// even containing newlines or curly braces. BE CAREFUL.
	public virtual void addAttribute(string st) {
		string key = "";
		string val = "";
		bool inQuotes = false;
		bool isVal = false;
		for(int i = 0; i < st.Length; i++) {
			// Some entity values in Source can use escape sequenced quotes. Need to make sure not to parse those.
			if(st[i]=='\"' && (i==0 || st[i-1]!='\\')) {
				if(inQuotes) {
					if(isVal) {
						break;
					}
					isVal=true;
				}
				inQuotes = !inQuotes;
			} else {
				if(inQuotes) {
					if(!isVal) {
						key+=st[i];
					} else {
						val+=st[i];
					}
				}
			}
		}
		val.Replace("\\\"", "\"");
		if(!string.IsNullOrWhiteSpace(key)) {
			if(!attributes.ContainsKey(key)) {
				attributes.Add(key, val);
			}
		}
	}

	public bool Remove(string target) {
		return attributes.Remove(target);
	}

	public void Add(string key, string val) {
		if(attributes.ContainsKey(key)) {
			attributes.Remove(key);
		}
		attributes.Add(key, val);
	}
	
	// +toString()
	// Returns the entity as an ASCII entity structure. The output of this method
	// reads as a complete entity that could be put into a map with no problems,
	// and that will be the primary use case for this method. Be sure to add the
	// newlines and curly braces around the entity if using this method to create
	// a new entities lump file, unless curly braces are part of the attributes
	// array.
	[Obsolete]
	public override string ToString() {
		string output = "entity {\n";
		foreach (string st in attributes.Keys) {
			output += "\""+st+"\" \""+attributes[st]+"\"\n";
		}
		foreach (MAPBrush brush in brushes) {
			output += "brush {\n"+brush.ToString()+"\n}\n";
		}
		return output+"}";
	}
	
	// attributeIs(String, String)
	// Returns true if the attribute String1 exists and is equivalent to String2
	public bool attributeIs(string attribute, string check) {
		return check.Equals(this[attribute], StringComparison.InvariantCultureIgnoreCase);
	}
	
	// Returns true if the bits in "spawnflags" corresponding to the set bits in 'check' are set
	public bool spawnflagsSet(int check) {
		return ((Spawnflags & check) == check);
	}
	
	// Toggles the bits in "spawnflags" which are set in "check"
	public void toggleSpawnflags(int toggle) {
		this["spawnflags"]=(Spawnflags ^ toggle).ToString();
	}
	
	// Disables the bits in "spawnflags" which are set in "check"
	// Alternate method: spawnflags = (disable ^ 0xFFFFFFFF) & spawnflags
	public void disableSpawnflags(int disable) {
		toggleSpawnflags(Spawnflags & disable);
	}
	
	// Enables the bits in "spawnflags" which are set in "check"
	public void enableSpawnflags(int enable) {
		this["spawnflags"]=(Spawnflags | enable).ToString();
	}
	
	public static Entity cloneNoBrushes(Entity copy) {
		return new Entity(copy.Attributes);
	}
	
	public virtual Entity cloneNoBrushes() {
		return new Entity(attributes);
	}

	public static Entity parseString(string data) {
		return new Entity(data.Split((char)0x0A));
	}
	
	public static Entities createLump(byte[] data) {
		int count = 0;
		bool inQuotes = false; // Keep track of whether or not we're currently in a set of quotation marks.
		// I came across a map where the idiot map maker used { and } within a value. This broke the code prior to revision 55.
		for (int i = 0; i < data.Length; i++) {
			if (inQuotes) {
				if (data[i] == '\"' && inQuotes) {
					inQuotes = false;
				}
			} else {
				if (data[i] == '\"') {
					inQuotes = true;
				} else {
					if (data[i] == '{') {
						count++;
					}
				}
			}
		}
		Entities lump = new Entities(data.Length, count);
		char currentChar; // The current character being read in the file. This is necessary because
		// we need to know exactly when the { and } characters occur and capture
		// all text between them.
		int offset = 0;
		for (int i = 0; i < count; i++) {
			// For every entity
			string current = ""; // This will be the resulting entity, fed into the Entity class
			currentChar = (char)data[offset]; // begin reading the file
			while (currentChar != '{') {
				// Eat bytes until we find the beginning of an entity structure
				offset++;
				currentChar = (char)data[offset];
			}
			inQuotes = false;
			do {
				if (currentChar == '\"') {
					inQuotes = !inQuotes;
				}
				current += (currentChar + ""); // adds characters to the current string
				offset++;
				currentChar = (char)data[offset];
			} while (currentChar != '}' || inQuotes); // Read bytes until we find the end of the current entity structure
			current += (currentChar + ""); // adds the '}' to the current string
			lump.Add(Entity.parseString(current));
		}
		return lump;
	}

	public byte[] toByteArray() {
		int len = 4;
		foreach(string key in attributes.Keys) {
			len += key.Length + attributes[key].Length + 6;
		}
		byte[] res = new byte[len];
		int offset = 0;
		res[offset++] = (byte)'{';
		res[offset++] = (byte)0x0A;
		foreach(string key in attributes.Keys) {
			res[offset++] = (byte)'\"'; // 1
			for(int i=0;i<key.Length;i++) {
				res[offset++] = (byte)key[i];
			}
			res[offset++] = (byte)'\"'; // 2
			res[offset++] = (byte)' '; // 3
			res[offset++] = (byte)'\"'; // 4
			for(int i=0;i<attributes[key].Length;i++) {
				res[offset++] = (byte)attributes[key][i];
			}
			res[offset++] = (byte)'\"'; // 5
			res[offset++] = (byte)0x0A; // 6
		}
		res[offset++] = (byte)'}';
		res[offset++] = (byte)0x0A;
		return res;
	}
	
	// ACCESSORS/MUTATORS
	virtual public bool BrushBased {
		get {
			return (brushes.Count > 0 || ModelNumber >= 0);
		}
	}

	virtual public int Spawnflags {
		get {
			try {
				return System.Int32.Parse(attributes["spawnflags"]);
			} catch {
				return 0;
			}
		}
	}

	virtual public bool Fired {
		get {
			return fired;
		}
		set {
			fired = value;
		}
	}

	virtual public Dictionary<string, string> Attributes {
		get {
			return attributes;
		}
	}

	virtual public List<MAPBrush> Brushes {
		get {
			return brushes;
		}
	}

	virtual public int ModelNumber {
		// If there's a model number in the attributes list, this method fetches it
		// and returns it. If there is no model defined, or it's not a numerical 
		// value, then -1 is returned. If it's the worldspawn then a 0 is returned.
		get {
			try {
				if(attributes["classname"] == "worldspawn") {
					return 0;
				} else {
					if(attributes.ContainsKey("model")) {
						string st = attributes["model"];
						if(st[0]=='*') {
							try {
								return System.Int32.Parse(st.Substring(1));
							} catch(System.FormatException) {
								return -1;
							}
						} else {
							return -1;
						}
					} else {
						return -1;
					}
				}
			} catch {
				return -1;
			}
		}
	}

	virtual public Vector3D Origin {
		// Returns the three components of the entity's "origin" attribute as a Vector3D
		get {
			double[] point = new double[3];
			if (attributes.ContainsKey("origin") && !string.IsNullOrWhiteSpace(attributes["origin"])) {
				string[] origin = attributes["origin"].Split(' ');
				for (int i = 0; i < 3 && i < origin.Length; i++) {
					try {
						point[i] = System.Double.Parse(origin[i]);
					} catch {
						return Vector3D.ZERO;
					}
				}
			}
			return new Vector3D(point);
		}
		set {
			this["origin"] = value.X+" "+value.Y+" "+value.Z;
		}
	}

	virtual public Vector3D Angles {
		// Returns the three components of the entity's "angles" attribute as a Vector3D
		get {
			double[] euler = new double[3];
			if (attributes.ContainsKey("angles") && !string.IsNullOrWhiteSpace(attributes["angles"])) {
				string[] angles = attributes["angles"].Split(' ');
				for (int i = 0; i < 3 && i < angles.Length; i++) {
					try {
						euler[i] = System.Double.Parse(angles[i]);
					} catch {
						return Vector3D.ZERO;
					}
				}
			}
			return new Vector3D(euler);
		}
		set {
			this["angles"] = value.X+" "+value.Y+" "+value.Z;
		}
	}
	
	// Set an attribute. If it doesn't exist, it is added. If it does, it is
	// overwritten with the new one, since that's much easier to do than edit
	// the preexisting one.
	public virtual string this[string attribute] {
		set {
			if(attributes.ContainsKey(attribute)) {
				attributes[attribute] = value;
			} else {
				attributes.Add(attribute, value);
			}
		}
		get {
			if(attributes.ContainsKey(attribute)) {
				return attributes[attribute];
			} else {
				return "";
			}
		}
	}
}